### PR TITLE
[APIM-4.0.0] Displaying comments with User Information.

### PR DESCRIPTION
**Pull request by Sarangan0219**
### Purpose

Fixes https://github.com/wso2/product-apim/issues/9595
 
This PR Provides the capability to view each commenter's information from the REST API. As the users are provisioned to APIM, we can use the default profile to extract the user's info. 

### Approach

We are introducing a new query parameter of **includeCommenterInfo** here. If this query parameter is provided in the request then commenter information with first name, last name, and Full name will be displayed. If not the commenter's Information will be null. The default value of this parameter is` false`.

For better performance, we have **userClaimsMap** which keeps track of the user information for a particular user. So we don't need to retrieve the claims from the user store for each comment. For a single user, even if there are multiple comments we need to query the user store only once.

**Update Resources:**

- GET --> /apis/{apiId}/comments
- GET -->  /apis/{apiId}/comments/{commentId}

1. **https://localhost:9443/api/am/store/v1/apis/95ed34cd-9afa-4f8c-a272-9fa8b41b16dd/comments?includeCommenterInfo=true**

```
{
    "count": 2,
    "list": [
        {
            "id": "6315a127-4eb5-4642-953a-1c7ae20ccac8",
            "content": "This is a comment from admin",
            "createdTime": 1606821504945,
            "createdBy": "admin",
            "commenterInfo": {
                "firstName": "Admin",
                "LastName": "Admin",
                "fullName": "Admin Admin"
            }
        },
        {
            "id": "79f91673-69e8-480d-9aa2-2094390cf02d",
            "content": "This is a comment from user saran",
            "createdTime": 1606821534430,
            "createdBy": "saran",
            "commenterInfo": {
                "firstName": "Saran",
                "LastName": "Jana",
                "fullName": "Saran Jana"
            }
        }
    ],
    "pagination": null
}
```

2. **https://localhost:9443/api/am/store/v1/apis/95ed34cd-9afa-4f8c-a272-9fa8b41b16dd/comments?includeCommenterInfo=false** 
 or  **https://localhost:9443/api/am/store/v1/apis/95ed34cd-9afa-4f8c-a272-9fa8b41b16dd/comments**

```
{
    "count": 2,
    "list": [
        {
            "id": "6315a127-4eb5-4642-953a-1c7ae20ccac8",
            "content": "This is a comment from admin",
            "createdTime": 1606821504945,
            "createdBy": "admin",
            "commenterInfo": null
        },
        {
            "id": "79f91673-69e8-480d-9aa2-2094390cf02d",
            "content": "This is a comment from user saran",
            "createdTime": 1606821534430,
            "createdBy": "saran",
            "commenterInfo": null
        }
    ],
    "pagination": null
}
```

3. **https://localhost:9443/api/am/store/v1/apis/95ed34cd-9afa-4f8c-a272-9fa8b41b16dd/comments/6315a127-4eb5-4642-953a-1c7ae20ccac8?includeCommenterInfo=true**

```
{
    "id": "79f91673-69e8-480d-9aa2-2094390cf02d",
    "content": "This is a comment from user saran",
    "createdTime": 1606821534430,
    "createdBy": "saran",
    "commenterInfo": {
        "firstName": "Saran",
        "LastName": "Jana",
        "fullName": "Saran Jana"
    }
}
```

4.  **https://localhost:9443/api/am/store/v1/apis/95ed34cd-9afa-4f8c-a272-9fa8b41b16dd/comments/6315a127-4eb5-4642-953a-1c7ae20ccac8** or **https://localhost:9443/api/am/store/v1/apis/95ed34cd-9afa-4f8c-a272-9fa8b41b16dd/comments/6315a127-4eb5-4642-953a-1c7ae20ccac8?includeCommenterInfo=false**

```
{
    "id": "6315a127-4eb5-4642-953a-1c7ae20ccac8",
    "content": "This is a comment from admin",
    "createdTime": 1606821504945,
    "createdBy": "admin",
    "commenterInfo": null
}
```

### Test Case Coverage

Screenshot from 2020-12-08 17-23-26

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml [hunk 1]
```diff
@@ -4590,6 +4593,18 @@ components:
             example: admin
           apiUUID:
             type: string
+    FullName:
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml:4596]:
```suggestion
    CommenterInfo:
```

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml [hunk 2]
```diff
@@ -4025,6 +4026,8 @@ components:
           type: string
           readOnly: true
           example: admin
+        commenterInformation:
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml:4029]:
```suggestion
        commenterInfo:
```

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml [hunk 3]
```diff
@@ -845,6 +845,7 @@ paths:
       - $ref: '#/components/parameters/requestedTenant'
       - $ref: '#/components/parameters/limit'
       - $ref: '#/components/parameters/offset'
+      - $ref: '#/components/parameters/commenterInfo'
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml:848]: shall we rename it as `includeCommenterInfo`

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 4]
```diff
@@ -2390,16 +2390,38 @@ public String getSubscriber(String subscriptionId) throws APIManagementException
...
+            if (subscriberClaims != null) {
+                for (String claimURI : configuredClaims.split(",")) {
+                    claimMap.put(claimURI, subscriberClaims.get(claimURI));
+                }
             }
+        } catch (UserStoreException e) {
+            throw new APIManagementException("Error while retrieving tenant id for tenant domain "
+                    + tenantDomain, e);
         }
+        return claimMap;
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:2407]: This is only a formatting fix right?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 5]
```diff
@@ -2390,16 +2390,38 @@ public String getSubscriber(String subscriptionId) throws APIManagementException
...
+     * @throws APIManagementException if failed to get user
+     */
+    @Override
+    public Map<String, String> getLoggedInUserClaims(String userName) throws APIManagementException {
+        String tenantDomain = MultitenantUtils.getTenantDomain(userName);
+        int tenantId = 0;
+        Map<String, String> claimMap;
+        try {
+            tenantId = getTenantId(tenantDomain);
+            claimMap = APIUtil.getClaims(userName, tenantId, ClaimsRetriever.DEFAULT_DIALECT_URI);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:2424]: I feel this method doesn't look right to have in APIProviderImpl.
Shall we move this method to APIUtil itself?

Comment (malinthaprasan) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:2424]: Ideal way to is to have these types of method in a separate Util class.. like APIMUserMgtUtils. please check with others as well.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml (hunk 1 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml]: https://github.com/wso2/carbon-apimgt/pull/9454/commits/219471b4eaadd05fd00560c3a5c735dedb45de6b

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml (hunk 2 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml]: https://github.com/wso2/carbon-apimgt/pull/9454/commits/219471b4eaadd05fd00560c3a5c735dedb45de6b

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml (hunk 3 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml]: https://github.com/wso2/carbon-apimgt/pull/9454/commits/219471b4eaadd05fd00560c3a5c735dedb45de6b

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 4 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:2407]: Yes

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 5 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:2424]: In the APIUtil class, we have the base claims retriever method(getClaims()). The use cases which are using that method should come from the APIProviderImpl or APIConsumerImpl nah?

Comment (malinthaprasan) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:2424]: It is not necessary. APIProvider and APIConsumer classes are actually APIM Domain specific..
APIProvider -> publisher operations
APIConsumer -> devportal operations
Retrieving user information is actually bit general and not specific to above two. So it is okay to have it in a Util class itself.

Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: According to our offline discussion created a new class **APIRealmUtil** to address this. Later we will port all user and role related methods from **APIUtil** class to this newly created class.

https://github.com/wso2/carbon-apimgt/pull/9454/commits/9dce063165691739762701864b94109afae195e2

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIRealmUtils.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
+package org.wso2.carbon.apimgt.impl.utils;
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIRealmUtils.java:1]: Please add license header

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/utils/APIRealmUtilTest.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
+package org.wso2.carbon.apimgt.impl.utils;
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/utils/APIRealmUtilTest.java:1]: License header

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml [hunk 8]
```diff
@@ -4590,6 +4594,18 @@ components:
             example: admin
           apiUUID:
             type: string
+    CommenterInfo:
+      type: object
+      properties:
+        firstName:
+          type: string
+          example: John
+        LastName:
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml:4603]:
```suggestion
        lastName:
```

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/mappings/CommentMappingUtil.java [hunk 9]
```diff
@@ -48,6 +53,43 @@ public static CommentDTO fromCommentToDTO(Comment comment) throws APIManagementE
...
+    /**
+     * Retrieve userClaims from UserStore and save it in a cache map.
+     *
+     * @param username        commenter username
+     * @param userClaimsMap   cache map with user deatils
+     * @return Map<String, Map<String, String>>
+     */
+    public static  Map<String, Map<String, String>> retrieveUserClaims(String username,
+                                                                       Map<String, Map<String, String>> userClaimsMap)
+            throws APIManagementException {
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/mappings/CommentMappingUtil.java:85]: Format the code

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java [hunk 10]
```diff
@@ -333,7 +333,14 @@ public Response getCommentOfAPI(String commentId, String apiId, String xWSO2Tena
             Comment comment = apiConsumer.getComment(identifier, commentId);
 
             if (comment != null) {
-                CommentDTO commentDTO = CommentMappingUtil.fromCommentToDTO(comment);
+                CommentDTO commentDTO;
+                if (includeCommenterInfo) {
+                    Map<String, Map<String, String>> userClaimsMap = new HashMap<>();
+                    CommentMappingUtil.retrieveUserClaims(comment.getUser(),userClaimsMap);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java:339]:
```suggestion
                    CommentMappingUtil.retrieveUserClaims(comment.getUser(), userClaimsMap);
```

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java [hunk 11]
```diff
@@ -44,17 +44,16 @@
...
 
 
 import java.io.File;
 import java.net.URI;
 import java.net.URISyntaxException;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java:56]: Do not use * import

Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java]: https://github.com/wso2/carbon-apimgt/pull/9454/commits/866313d7b897152f88b7c3225827d8df4587365a

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java (hunk 10 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java]: https://github.com/wso2/carbon-apimgt/pull/9454/commits/866313d7b897152f88b7c3225827d8df4587365a

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml (hunk 8 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/resources/store-api.yaml]: https://github.com/wso2/carbon-apimgt/pull/9454/commits/866313d7b897152f88b7c3225827d8df4587365a

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/utils/APIRealmUtilTest.java (hunk 7 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/utils/APIRealmUtilTest.java:18]: https://github.com/wso2/carbon-apimgt/pull/9454/commits/866313d7b897152f88b7c3225827d8df4587365a

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIRealmUtils.java (hunk 6 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIRealmUtils.java:19]: https://github.com/wso2/carbon-apimgt/pull/9454/commits/866313d7b897152f88b7c3225827d8df4587365a

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/mappings/CommentMappingUtil.java (hunk 9 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/mappings/CommentMappingUtil.java:85]: We can't include `Map<String, Map<String, String>> userClaimsMap` in line 83. It is exceeding the number of characters limit

Comment (malinthaprasan) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/mappings/CommentMappingUtil.java:85]: can we merge 85 and 84 to a single line?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java [hunk 12]
```diff
@@ -333,7 +334,14 @@ public Response getCommentOfAPI(String commentId, String apiId, String xWSO2Tena
             Comment comment = apiConsumer.getComment(identifier, commentId);
 
             if (comment != null) {
-                CommentDTO commentDTO = CommentMappingUtil.fromCommentToDTO(comment);
+                CommentDTO commentDTO;
+                if (includeCommenterInfo) {
+                    Map<String, Map<String, String>> userClaimsMap = new HashMap<>();
+                    CommentMappingUtil.retrieveUserClaims(comment.getUser(), userClaimsMap);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java:340]:
```suggestion
                    Map<String, Map<String, String>> userClaimsMap = CommentMappingUtil.retrieveUserClaims(comment.getUser(), new HashMap<>());
```
Will this be more readable?

Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApisApiServiceImpl.java]: https://github.com/wso2/carbon-apimgt/pull/9454/commits/dbe6f847730212f0db8456b3628c917d31a75961

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/mappings/CommentMappingUtil.java (hunk 9 above)
Comment (Sarangan0219) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/mappings/CommentMappingUtil.java]: https://github.com/wso2/carbon-apimgt/pull/9454/commits/dbe6f847730212f0db8456b3628c917d31a75961

**Review by malinthaprasan — APPROVED** [review]

**Pull request merged**